pkg/controller/v1beta1/inferenceservice/components/predictor.go

```diff
@@ -198,51 +203,103 @@ func (p *Predictor) Reconcile(ctx context.Context, isvc *v1beta1.InferenceServic
+
+		containerIndexInSR := isvcutils.GetContainerIndexByName(sRuntime.Containers, constants.TransformerContainerName)
+		containerIndexInIS := isvcutils.GetContainerIndexByName(isvc.Spec.Predictor.Containers, constants.TransformerContainerName)
+		if containerIndexInSR != -1 {
+			// Replace placeholders in transformer container by values from inferenceservice metadata
+			if err = isvcutils.ReplacePlaceholders(&sRuntime.Containers[containerIndexInSR], isvc.ObjectMeta); err != nil {
+				isvc.Status.UpdateModelTransitionStatus(v1beta1.InvalidSpec, &v1beta1.FailureInfo{
+					Reason:  v1beta1.InvalidPredictorSpec,
+					Message: fmt.Sprintf(ErrInvalidPlaceholder, *isvc.Spec.Predictor.Model.Runtime, sRuntime.Containers[containerIndexInSR].Name),
+				})
+				return ctrl.Result{}, errors.Wrapf(err, ErrInvalidPlaceholder, *isvc.Spec.Predictor.Model.Runtime, sRuntime.Containers[containerIndexInSR].Name)
+			}
+		}
+		switch {
+		// Transformer container is present in both ServingRuntime and InferenceService
+		case containerIndexInIS != -1 && containerIndexInSR != -1:
+			// Merge transformer container from ServingRuntime and InferenceService
+			transformerContainer, err := isvcutils.MergeRuntimeContainers(&sRuntime.Containers[containerIndexInSR], &isvc.Spec.Predictor.Containers[containerIndexInIS])
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			podSpec.Containers = append(podSpec.Containers, *transformerContainer)
+			// Append all containers except the transformer container from InferenceService
+			podSpec.Containers = append(podSpec.Containers, isvc.Spec.Predictor.Containers[:containerIndexInIS]...)
+			podSpec.Containers = append(podSpec.Containers, isvc.Spec.Predictor.Containers[containerIndexInIS+1:]...)
+			// Append all containers except the predictor container and transformer container from ServingRuntime
+			for _, container := range sRuntime.Containers {
+				if container.Name != constants.InferenceServiceContainerName && container.Name != constants.TransformerContainerName {
+					podSpec.Containers = append(podSpec.Containers, container)
+				}
+			}
+		// Transformer container is present in InferenceService
+		case containerIndexInIS != -1:
+			// Append transformer container from InferenceService
+			podSpec.Containers = append(podSpec.Containers, isvc.Spec.Predictor.Containers[containerIndexInIS])
+			// Append all containers except the transformer container from InferenceService
+			podSpec.Containers = append(podSpec.Containers, isvc.Spec.Predictor.Containers[:containerIndexInIS]...)
+			podSpec.Containers = append(podSpec.Containers, isvc.Spec.Predictor.Containers[containerIndexInIS+1:]...)
+			// Append all containers except predictor container from ServingRuntime
+			podSpec.Containers = append(podSpec.Containers, sRuntime.Containers[:kserveContainerIdx]...)
+			podSpec.Containers = append(podSpec.Containers, sRuntime.Containers[kserveContainerIdx+1:]...)
```

Comment:
We can simplify this logic across all the cases by excluding the kserve and transformer container from both serving runtime and isvc, then add the rest of the containers.